superset/views/sql_lab.py

```diff
@@ -169,6 +175,157 @@ def show(self, pk):
 appbuilder.add_view_no_menu(SavedQueryViewApi)
 appbuilder.add_view_no_menu(SavedQueryView)
 
+
+class TabStateView(BaseSupersetView):
+    def _get_owner_id(self, tab_state_id):
+        return db.session.query(TabState.user_id).filter_by(id=tab_state_id).scalar()
+
+    @has_access_api
+    @expose("/", methods=["POST"])
+    def post(self):
+        query_editor = json.loads(request.form["queryEditor"])
+        if query_editor.get("latestQueryId"):
+            query = (
+                db.session.query(Query)
+                .filter_by(client_id=query_editor.get("latestQueryId"))
+                .scalar()
+            )
+        else:
+            query = Query(
+                client_id=utils.shortid()[:10],
+                database_id=query_editor["dbId"],
+                schema=query_editor.get("schema"),
+                sql=query_editor.get("sql", "SELECT ..."),
+            )
+        tab_state = TabState(
+            user_id=g.user.get_id(),
+            label=query_editor.get("title", "Untitled Query"),
+            active=True,
+            database_id=query_editor["dbId"],
+            schema=query_editor.get("schema"),
+            query=query,
+            query_limit=query_editor.get("queryLimit"),
+        )
+        (
+            db.session.query(TabState)
+            .filter_by(user_id=g.user.get_id())
+            .update({"active": False})
+        )
+        db.session.add(tab_state)
+        db.session.commit()
+        return json_success(json.dumps({"id": tab_state.id}))
+
+    @has_access_api
+    @expose("/<int:tab_state_id>", methods=["DELETE"])
+    def delete(self, tab_state_id):
+        db.session.query(TabState).filter(TabState.id == tab_state_id).delete(
```

Comment:
does `has_access_api` do the validation that the user making the request owns this tab?

Reply:
@dpgaspar can chime in here, but I'm pretty sure it does not.

Reply:
Gotcha, we definitely want to add that check then